subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/AbstractCacheCleanup.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.cache.internal;
+
+import org.apache.commons.io.FileUtils;
+import org.gradle.api.Action;
+import org.gradle.cache.PersistentCache;
+import org.gradle.internal.operations.BuildOperationContext;
+import org.gradle.internal.operations.BuildOperationExecutor;
+import org.gradle.internal.operations.CallableBuildOperation;
+import org.gradle.internal.operations.RunnableBuildOperation;
+import org.gradle.internal.progress.BuildOperationDescriptor;
+import org.gradle.util.GFileUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.FilenameFilter;
+import java.util.List;
+
+abstract class AbstractCacheCleanup implements Action<PersistentCache> {
```

Comment:
This looks like something that should be applied as a decoration over some new 'cache cleanup action' type, rather than inherited from to present `Action<PersistentCache>`. This way, for example, we don't need to expose the services required by this general purpose code in the constructors of the specific cleanup action. We'd replace the `Action` parameter of `CacheBuilder.withCleanup(Action<? super PersistentCache> cleanup)` with the new type.

Reply:
I started down this path, which pulled `BuildOperationExecutor` into the `CacheBuilder` dependencies.  This seems to be used in a few places that don't have a real `BuildOperationExecutor` available (e.g., the Zinc compiler daemon).  I backed off on the decoration.  I still have a new `CleanupAction`, but the decoration is done in a `CleanupActionFactory` instead. This moves the `BuildOperationExecutor` dependency to the factory and I can inject the factory into the two places that need to create `CleanupAction`s.